### PR TITLE
Filter ecs task app cloudwatch log stream based on time

### PR DESCRIPTION
**Pull request by chrisgilmerproj**
## Description

For #2691 - this change allows us to filter based on time stamps. The implementation isn't done but it gets the point across. Here's how you'd use it:

```sh
$ ecs-service-logs show -s app -e experimental event_type="created" --status=ALL --start-time=2019-09-16T23:43:00Z --end-time=2019-09-16T23:43:20Z
{"level":"info","ts":"2019-09-16T23:43:10.562Z","caller":"adminapi/office_users.go:122","msg":"Create Office User","git_branch":"gm-ig-log-auditing","git_commit":"8ee3b5fca5089c211c1bb14d1571f4791ba092c0","ecs_cluster":"arn:aws:ecs:us-west-2:923914045601:cluster/app-experimental","ecs_task_def_family":"app-experimental","ecs_task_def_revision":"393","milmove_trace_id":"0cbd4167-7dcd-4300-b304-2eed8499695b","office_user_id":"f25f9a15-d4d1-4ba1-a84b-5ae7e33d12c4","responsible_user_id":"1ac2926b-aa8c-4169-9ba7-be6183a96eee","event_type":"created"}
```

### CODE REVIEW

File: cmd/ecs-service-logs/main.go [hunk 1]
```diff
@@ -171,7 +173,7 @@ func initFlags(flag *pflag.FlagSet) {
 	flag.StringP(flagTaskDefinitionRevision, "r", "", "The ECS task definition revision.")
 	flag.IntP(flagPageSize, "p", -1, "The page size or maximum number of log events to return during each API call.  The default is 10,000 log events.")
 	flag.IntP(flagLimit, "n", -1, "If 1 or above, the maximum number of log events to print to stdout.")
-	flag.IntP(flagTasks, "t", 10, "If 1 or above, the maximum number of log streams (aka tasks) to print to stdout.")
+	flag.IntP(flagTasks, "t", 1000, "If 1 or above, the maximum number of log streams (aka tasks) to print to stdout.")
```
Comment (chrisgilmerproj) [inline cmd/ecs-service-logs/main.go]: Its not really necessary to do this but I wanted to not have it block me. If we filter based on time ranges then probably 10 is enough.

**Comment by garrettqmartin8** [comment]
@chrisgilmerproj meant to make the above a comment, not an approval :)

File: cmd/ecs-service-logs/main.go [hunk 2]
```diff
@@ -425,6 +446,28 @@ func showFunction(cmd *cobra.Command, args []string) error {
...
+
+		endTime, errEndTime := time.Parse(time.RFC3339, endTimeString)
+		if errEndTime != nil {
+			return errEndTime
+		}
+
+		startTimeUnixTmp := startTime.Unix() * 1000 // milliseconds
+		endTimeUnixTmp := endTime.Unix() * 1000     // milliseconds
+		startTimeUnix = &startTimeUnixTmp
+		endTimeUnix = &endTimeUnixTmp
```
Comment (chrisgilmerproj) [inline cmd/ecs-service-logs/main.go]: This seems janky. Maybe I ought to use `aws.Int64()` here.

File: cmd/ecs-service-logs/main.go [hunk 3]
```diff
@@ -731,6 +783,10 @@ func showFunction(cmd *cobra.Command, args []string) error {
 				LogStreamNames: []*string{aws.String(job.LogStreamName)},
 				NextToken:      nextToken,
 			}
+			if startTimeUnix != nil && endTimeUnix != nil {
+				filterLogEventsInput.StartTime = startTimeUnix
+				filterLogEventsInput.EndTime = endTimeUnix
```
Comment (chrisgilmerproj) [inline cmd/ecs-service-logs/main.go]: If these are `nil` then can I just assign them inside the struct above?

**Comment by chrisgilmerproj** [comment]
> This looks great, how can I help get it over the finish line?

What this needs is some testing and cleanup. I've left comments about the cleanup that could probably be done. And if you can run the tool without the timestamps and with the timestamps then that's all the testing we need I think. Would also be good to check that if you leave out a timestamp or reverse the order of the timestamps that it fails to work and gives an error message.

Do you have the bandwidth to take this on or would you like me to work on finishing it?

**Comment by garrettqmartin8** [comment]
I probably won't have the bandwidth this week as I'm going to be out on Friday, but I can take a look next week if you don't mind waiting.

**Comment by chrisgilmerproj** [comment]
I don't mind waiting as long as its done by the end of next week. I would hate to sit on this longer than that.

**Comment by chrisgilmerproj** [comment]
@garrettqmartin8 - any update on this? Did you want to pair on it?

**Comment by garrettqmartin8** [comment]
@chrisgilmerproj not yet, sorry. I had some time-sensitive GHC work that landed on my plate this past week. I'll look into this in the next sprint starting today.

**Comment by garrettqmartin8** [comment]
@chrisgilmerproj - I did some cleanup based on your comments and added an error to be thrown when the timestamps are backwards. Tested with and without the time flags, so this is looking pretty good. Let me know what you think.

File: cmd/ecs-service-logs/main.go [hunk 4]
```diff
@@ -425,6 +446,28 @@ func showFunction(cmd *cobra.Command, args []string) error {
...
+		endTime, errEndTime := time.Parse(time.RFC3339, endTimeString)
+		if errEndTime != nil {
+			return errEndTime
+		}
+
+		startTimeUnix = aws.Int64(startTime.Unix() * 1000) // milliseconds
+		endTimeUnix = aws.Int64(endTime.Unix() * 1000)     // milliseconds
+
+		if *startTimeUnix > *endTimeUnix {
+			return errors.New("the time ranges you specified do not overlap")
```
Comment (chrisgilmerproj) [inline cmd/ecs-service-logs/main.go]: I'd use the special error here.

```suggestion
			return &errInvalidTimeRange{StartTime: startTime, EndTime: endTime}
```

**Comment by chrisgilmerproj** [comment]
@garrettqmartin8 - thank you for doing the work on this! I have one requested change. Then I say we get this merged :)

**Comment by chrisgilmerproj** [comment]
@garrettqmartin8 - looks like you have it all fixed. Sorry the server tests are flaky but when they pass go ahead and merge:)

**Pull request merged**